### PR TITLE
Golden rules for Magic Modules development

### PR DESCRIPTION
**Pull request by nelsonjr**
List of best practices to be observed when developing Magic Modules code.

### CODE REVIEW

File: GOVERNANCE.md [hunk 1]
```diff
@@ -0,0 +1,413 @@
+# Magic Modules Governance (a.k.a. "Golden Rules")
+
```
Comment (erjohnso) [inline GOVERNANCE.md:2]: As a preamble, I'd suggest something like,

> Please refer to the top-level [README](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/README.md) and [Philosophy](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/docs/philosophy.md) before reading this guide. All code-reviews of Magic Modules (regardless of changes to core, providers, or products) must follow the guidelines outlined here.

File: GOVERNANCE.md [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+have a complicated logic (e.g. 'build a map between URL and properties') it is
+best to put this functionality in the `provider/<provider>.rb` class.  All
+methods in the provider class are directly available to the template (e.g. the
+`indent()` function).
+
+If the function is useful for all providers then consider adding it to the
+global (core) provider namespace.
+
+
+## Testing
```
Comment (erjohnso) [inline GOVERNANCE.md:94]: I'd call this "Development Principles" or something like that. I also suggest making this the first section, preceding the folder/location section.

File: GOVERNANCE.md [hunk 3]
```diff
@@ -0,0 +1,413 @@
...
+      * provider independent
+  - `provider/<provider>/...`: Holds all provider specific code
+      * product independent
+  - `templates`: Holds all global templates
+      * product independent
+      * provider independent
+  - `templates/<provider>/...`: Holds all templates for the provider
+      * product independent
+
+In the list above folders without `...` means the folder only. `...` represents
```
Comment (erjohnso) [inline GOVERNANCE.md:47]: Move this and the 'Definitions' section directly under the 'folder/location' heading. The actual folder list should be presented after this intro material.

> In the list below, folders without the `...` means...

File: GOVERNANCE.md [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+  - Code coverage **must** always stay >80% and **should** be >90%
+
+Refer to [CONTRIBUTING][contrib] for details on how to test your changes.
+
+
+## Exceptions
+
+Any _permanent_ exceptions to these rules **must** be thoroughly documented in
+the code. If a longer discussion is required and becomes beyond the code where
+it lives an issue **must** be created and referenced for context.
```
Comment (erjohnso) [inline GOVERNANCE.md:115]: I'm assuming the issue also has to have LGTMs on it from other core team members.

**Review by erjohnso — COMMENTED** [review]
Thanks for putting this together. Mostly suggestions around ordering.

File: GOVERNANCE.md (hunk 1 above)
Comment (nelsonjr) [reply GOVERNANCE.md:2]: Should this be instead of the first paragraph, or in addition to?

File: GOVERNANCE.md (hunk 2 above)
Comment (nelsonjr) [reply GOVERNANCE.md:94]: Done.

File: GOVERNANCE.md (hunk 3 above)
Comment (nelsonjr) [reply GOVERNANCE.md:47]: Done.

File: GOVERNANCE.md (hunk 4 above)
Comment (nelsonjr) [reply GOVERNANCE.md:115]: Done.

**Pull request merged**